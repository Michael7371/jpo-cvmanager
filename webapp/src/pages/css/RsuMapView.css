.dateStyle {
  height: fit-content;
  width: fit-content;
  position: absolute;
  right: 50px;
  top: 160px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 30px;
  padding-right: 30px;
  font-size: 1em;
  background: white;
  border-radius: 10px;
  color: black;
}

.backButton {
  height: fit-content;
  width: fit-content;
  position: absolute;
  left: 50px;
  top: 160px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 30px;
  padding-right: 30px;
  font-size: 1em;
  background: #d16d15;
  border-radius: 20px;
  border-color: white;
  cursor: pointer;
  font-weight: bold;
  color: white;
}

.srmSsmToggle {
  height: fit-content;
  width: fit-content;
  position: absolute;
  left: 50px;
  top: 220px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 30px;
  padding-right: 30px;
  font-size: 1em;
  background: #d16d15;
  border-radius: 20px;
  border-color: white;
  cursor: pointer;
  font-weight: bold;
  color: white;
}

#toggle {
  border: none;
  padding: 1px 10px;
  text-align: center;
  text-decoration: none;
  font-size: 16px;
  margin: 4px 2px;
  cursor: pointer;
  border-radius: 15px;
  margin-left: 100px;
  position: absolute;
  z-index: 100;
  margin-top: 10px;
  right: 10px;
}

.ssmSrmContainer {
  height: calc(100vh - 200px);
  width: 500px;
  position: absolute;
  top: 220px;
  left: 0px;
  background: white;
  overflow-y: scroll;
  text-align: center;
}

#ssmsrmDataHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bolder;
  margin: 15px;
  margin-top: 25px;
}

#ssmSrmHeaderContainer {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  text-align: center;
  padding: 10px;
  border-bottom: 1px #0e2052 solid;
}

#countsContainer {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  text-align: center;
  margin-top: 15px;
  margin-left: 10px;
  margin-right: 10px;
}

#countsHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bolder;
  margin: 10px;
}

#countsData {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bold;
  width: 90px;
  padding: 5px;
  margin-bottom: 20px;
}

#timeHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bold;
  width: 90px;
  margin-left: 10px;
}

#requestHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bold;
  width: 100px;
  margin-left: 20px;
}

#roleHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bold;
  width: 90px;
  margin-left: 40px;
}

#statusHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bold;
  width: 90px;
  padding-left: 100px;
}

#ssmSrmHeader {
  font-family: Arial, Helvetica, sans-serif;
  color: #0e2052;
  font-weight: bold;
  width: 100px;
}

.keyContainer {
  height: fit-content;
  width: 180px;
  position: absolute;
  background-color: white;
  border-radius: 10px;
  right: 50px;
  top: 225px;
  padding: 15px 5px;
}

#referenceWrapper {
  display: flex;
  height: fit-content;
  justify-content: flex-start;
  align-items: center;
  width: 90%;
}

#ingressKey {
  display: inline-block;
  height: 20px;
  width: 20px;
  background-color: rgb(50, 205, 50);
  margin: 8px 5px;
  border: 0.5px solid black;
  border-radius: 10px;
}

.referenceInfo {
  display: inline-block;
  font-size: 18px;
}

#keyTitle {
  width: 100%;
  text-align: center;
  margin: 0px 0px 10px 0px;
}

#egressKey {
  display: inline-block;
  background-color: rgb(203, 4, 4);
  height: 20px;
  width: 20px;
  margin: 8px 5px;
  border: 0.5px solid black;
  border-radius: 10px;
}

#ssmKey {
  display: inline-block;
  background-color: #fff1d3;
  height: 20px;
  width: 20px;
  margin: 8px 5px;
  border: 0.5px solid black;
  border-radius: 10px;
}

#srmKey {
  display: inline-block;
  background-color: #bdceff;
  height: 20px;
  width: 20px;
  margin: 8px 5px;
  border: 0.5px solid black;
  border-radius: 10px;
}

#srmKeyMap {
  display: inline-block;
  background-color: #0e2052;
  height: 20px;
  width: 20px;
  margin: 8px 5px;
  border: 0.5px solid black;
  border-radius: 10px;
}
