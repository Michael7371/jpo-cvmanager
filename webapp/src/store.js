import { configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import rsuUpdateMenuReducer from './features/rsuUpdateMenu/rsuUpdateMenuSlice'

export const setupStore = (preloadedState) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      rsuUpdateMenu: rsuUpdateMenuReducer,
    },
    preloadedState,
  })
}
