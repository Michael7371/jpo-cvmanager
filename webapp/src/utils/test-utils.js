/**
 * This function searches for the every somewhat random/chaotic class name and property that cause snapshot tests to be inconsistent.
 * This current list includes MUI classes (css-*) and aria-invalid attributes.
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceChaoticIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container) {
  const props = [
    {
      selector: 'class',
      updateFunc: (val) => val.replace(/css-[0-9a-z]*?-/g, 'css-mocked-'),
    },
  ]

  container.querySelectorAll('*').forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)))
      }
    })
  })
  container.querySelectorAll('input[aria-invalid]').forEach((item) => {
    item.removeAttribute('aria-invalid')
  })
  return container
}

export { replaceChaoticIds }
