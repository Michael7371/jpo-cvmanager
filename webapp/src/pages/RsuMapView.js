import React, { useEffect, useState } from 'react'
import Map, { Source, Layer } from 'react-map-gl'
import { Container } from 'reactstrap'
import EnvironmentVars from '../EnvironmentVars'
import './css/RsuMapView.css'
import SsmSrmItem from '../components/SsmSrmItem'
import { useSelector, useDispatch } from 'react-redux'
import {
  selectRsuMapData,
  selectSelectedRsu,
  selectSelectedSrm,
  selectMapDate,
  selectSsmDisplay,
  selectRsuIpv4,
  selectSrmSsmList,

  // actions
  toggleMapDisplay,
  toggleSsmSrmDisplay,
  getSsmSrmData,
} from '../generalSlices/rsuSlice'

function RsuMapView(props) {
  const dispatch = useDispatch()

  const rsuMapData = useSelector(selectRsuMapData)
  const selectedRsu = useSelector(selectSelectedRsu)
  const selectedSrm = useSelector(selectSelectedSrm)
  const mapDate = useSelector(selectMapDate)
  const ssmDisplay = useSelector(selectSsmDisplay)
  const rsuIpv4 = useSelector(selectRsuIpv4)
  const srmSsmList = useSelector(selectSrmSsmList)

  const [srmCount, setSrmCount] = useState(0)
  const [ssmCount, setSsmCount] = useState(0)
  const [msgList, setMsgList] = useState([])
  const [egressData, setEgressData] = useState({
    type: 'FeatureCollection',
    features: [],
  })
  const [ingressData, setIngressData] = useState({
    type: 'FeatureCollection',
    features: [],
  })

  useEffect(() => {
    dispatch(getSsmSrmData())
  }, [dispatch])

  useEffect(() => {
    let localSrmCount = 0
    let localSsmCount = 0
    let localMsgList = []
    for (const elem of srmSsmList) {
      if (elem.ip === rsuIpv4) {
        localMsgList.push(elem)
        if (elem.type === 'srmTx') {
          localSrmCount += 1
        } else {
          localSsmCount += 1
        }
      }
    }
    setSrmCount(localSrmCount)
    setSsmCount(localSsmCount)
    setMsgList(localMsgList)
  }, [srmSsmList, rsuIpv4])

  useEffect(() => {
    const ingressDataFeatures = []
    const egressDataFeatures = []

    Object.entries(rsuMapData?.['features'] ?? []).map((feature) => {
      if (feature[1].properties.ingressPath === 'true') {
        ingressDataFeatures.push(feature[1])
      }
      return null
    })
    Object.entries(rsuMapData?.['features'] ?? []).map((feature) => {
      if (feature[1].properties.egressPath === 'true') {
        egressDataFeatures.push(feature[1])
      }
      return null
    })

    setIngressData((prevIngressData) => {
      return { ...prevIngressData, features: ingressDataFeatures }
    })
    setEgressData((prevEgressData) => {
      return { ...prevEgressData, features: egressDataFeatures }
    })
  }, [rsuMapData])

  const srmData = {
    type: 'FeatureCollection',
    features: [],
  }

  if (selectedSrm.length > 0) {
    srmData.features.push({
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [selectedSrm[0].long, selectedSrm[0].lat],
      },
    })
  }

  const ingressLayer = {
    id: 'ingressLayer',
    type: 'line',
    minzoom: 14,
    source: 'ingressData',
    layout: {
      'line-join': 'round',
      'line-cap': 'round',
    },
    paint: {
      'line-color': 'rgb(50,205,50)',
      'line-width': 3,
    },
  }

  const egressLayer = {
    id: 'egressLayer',
    type: 'line',
    minzoom: 14,
    source: 'egressData',
    layout: {
      'line-join': 'round',
      'line-cap': 'round',
    },
    paint: {
      'line-color': 'rgb(203, 4, 4)',
      'line-width': 3,
    },
  }

  const srmLayer = {
    id: 'srmMarker',
    type: 'circle',
    source: 'srmData',
    minzoom: 12,
    paint: {
      'circle-radius': 8,
      'circle-color': 'rgb(14, 32, 82)',
    },
  }

  const [viewState, setViewState] = useState({
    latitude: selectedRsu.geometry.coordinates[1],
    longitude: selectedRsu.geometry.coordinates[0],
    zoom: 17,
  })

  return (
    <div className="container">
      <Container
        fluid={true}
        style={{ width: '100%', height: props.auth ? 'calc(100vh - 135px)' : 'calc(100vh - 100px)', display: 'flex' }}
      >
        <Map
          {...viewState}
          mapboxAccessToken={EnvironmentVars.MAPBOX_TOKEN}
          mapStyle={'mapbox://styles/mapbox/satellite-v9'}
          style={{ width: '100%', height: '100%' }}
          onMove={(evt) => {
            setViewState(evt.viewState)
          }}
        >
          <Source type="geojson" data={ingressData}>
            <Layer {...ingressLayer} />
          </Source>
          <Source type="geojson" data={egressData}>
            <Layer {...egressLayer} />
          </Source>
          <Source type="geojson" data={srmData}>
            <Layer {...srmLayer} />
          </Source>
        </Map>
      </Container>
      <button className="backButton" onClick={(e) => dispatch(toggleMapDisplay())}>
        Back
      </button>
      <div className="dateStyle">MAP data from {mapDate}</div>
      {ssmDisplay ? (
        <div className="ssmSrmContainer">
          <button id="toggle" onClick={() => dispatch(toggleSsmSrmDisplay())}>
            X
          </button>
          <h3 id="ssmsrmDataHeader">SSM / SRM Data For {rsuIpv4}</h3>
          <div id="ssmSrmHeaderContainer">
            <p id="timeHeader"> Time </p>
            <p id="requestHeader"> Request Id</p>
            <p id="roleHeader"> Role </p>
            <p id="ssmSrmHeader"> Status </p>
            <p id="ssmSrmHeader"> Display </p>
          </div>
          {Object.keys(msgList).map((index) => (
            <SsmSrmItem key={index} index={index} elem={msgList[index]} setSelectedSrm={selectedSrm} />
          ))}
          <h3 id="countsHeader"> Total Counts </h3>
          <div id="countsContainer">
            <h4 id="countsData"> SSM: {ssmCount} </h4>
            <h4 id="countsData"> SRM: {srmCount} </h4>
          </div>
        </div>
      ) : (
        <button className="srmSsmToggle" onClick={() => dispatch(toggleSsmSrmDisplay())}>
          SSM/SRM Display
        </button>
      )}
      <div className="keyContainer">
        <div id="keyTitle">
          <h3 id="keyTitle">Reference</h3>
        </div>
        <div id="referenceWrapper">
          <div id="ingressKey" />
          <p className="mapKeyInfo">- Ingress Lane</p>
        </div>
        <div id="referenceWrapper">
          <div id="egressKey" />
          <p className="mapKeyInfo">- Egress Lane</p>
        </div>
        <div id="referenceWrapper">
          <div id="ssmKey" />
          <p className="referenceInfo">- SSM (table)</p>
        </div>
        <div id="referenceWrapper">
          <div id="srmKey" />
          <p className="referenceInfo">- SRM (table)</p>
        </div>
        <div id="referenceWrapper">
          <div id="srmKeyMap" />
          <p className="referenceInfo">- SRM (map)</p>
        </div>
      </div>
    </div>
  )
}

export default RsuMapView
