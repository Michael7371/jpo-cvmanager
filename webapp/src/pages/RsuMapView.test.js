import React from 'react'
import { render } from '@testing-library/react'
import RsuMapView from './RsuMapView'
import { Provider } from 'react-redux'
import { setupStore } from '../store'
import { replaceChaoticIds } from '../utils/test-utils'

it('should take a snapshot', () => {
  const initialState = {
    rsu: { value: { selectedRsu: { geometry: { coordinates: [0, 1] } }, selectedSrm: [], srmSsmList: [] } },
  }
  const { container } = render(
    <Provider store={setupStore(initialState)}>
      <RsuMapView />
    </Provider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
