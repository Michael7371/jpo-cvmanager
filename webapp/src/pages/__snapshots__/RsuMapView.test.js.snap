// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="container-fluid"
      style="width: 100%; height: calc(100vh - 100px); display: flex;"
    >
      <div
        style="position: relative; width: 100%; height: 100%;"
      />
    </div>
    <button
      class="backButton"
    >
      Back
    </button>
    <div
      class="dateStyle"
    >
      MAP data from 
    </div>
    <button
      class="srmSsmToggle"
    >
      SSM/SRM Display
    </button>
    <div
      class="keyContainer"
    >
      <div
        id="keyTitle"
      >
        <h3
          id="keyTitle"
        >
          Reference
        </h3>
      </div>
      <div
        id="referenceWrapper"
      >
        <div
          id="ingressKey"
        />
        <p
          class="mapKeyInfo"
        >
          - Ingress Lane
        </p>
      </div>
      <div
        id="referenceWrapper"
      >
        <div
          id="egressKey"
        />
        <p
          class="mapKeyInfo"
        >
          - Egress Lane
        </p>
      </div>
      <div
        id="referenceWrapper"
      >
        <div
          id="ssmKey"
        />
        <p
          class="referenceInfo"
        >
          - SSM (table)
        </p>
      </div>
      <div
        id="referenceWrapper"
      >
        <div
          id="srmKey"
        />
        <p
          class="referenceInfo"
        >
          - SRM (table)
        </p>
      </div>
      <div
        id="referenceWrapper"
      >
        <div
          id="srmKeyMap"
        />
        <p
          class="referenceInfo"
        >
          - SRM (map)
        </p>
      </div>
    </div>
  </div>
</div>
`;
